source/common/buffer/buffer_impl.h

```diff
@@ -30,22 +30,67 @@ namespace Buffer {
  *                   |
  *                   data()
  */
-class Slice : public SliceData {
+class Slice {
 public:
   using Reservation = RawSlice;
 
-  ~Slice() override { callAndClearDrainTrackers(); }
+  Slice() : storage_(nullptr), base_(nullptr), data_(0), reservable_(0), capacity_(0) {}
 
-  // SliceData
-  absl::Span<uint8_t> getMutableData() override {
-    RELEASE_ASSERT(isMutable(), "Not allowed to call getMutableData if slice is immutable");
-    return {base_ + data_, static_cast<absl::Span<uint8_t>::size_type>(reservable_ - data_)};
+  Slice(uint64_t size)
+      : storage_(new uint8_t[size]), base_(storage_.get()), data_(0), reservable_(0),
+        capacity_(size) {}
+
+  Slice(BufferFragment& fragment)
+      : storage_(nullptr), base_(static_cast<uint8_t*>(const_cast<void*>(fragment.data()))),
+        data_(0), reservable_(fragment.size()), capacity_(fragment.size()) {
+    addDrainTracker([&fragment]() { fragment.done(); });
+  }
+
+  Slice(Slice&& rhs) noexcept {
+    storage_ = std::move(rhs.storage_);
```

Comment:
TODO: Use helper function to code shared between move constructor and move assignment?